run_training.py

```diff
@@ -0,0 +1,1399 @@
+        if not time_matching_mat is None:
+            z_ = z.reshape((z.shape[0], -1))
+            len_latent = z_.shape[1]
+            sim_mat = t.pow(z_.reshape((1, -1, len_latent)) - \
+                            z_.reshape((-1, 1, len_latent)), 2).mean(2)
+            assert sim_mat.shape == time_matching_mat.shape
+            time_matching_loss = (sim_mat * time_matching_mat).sum()
+            total_loss += time_matching_loss * self.alpha
+        return decoded, \
+               {'recon_loss': recon_loss,
+                'time_matching_loss': time_matching_loss,
+                'total_loss': total_loss,
+                'perplexity': t.zeros(())}
+
+    def adversarial_loss(self, inputs):
+        """ Calculate adversarial loss for the batch
+
+        Args:
+            inputs (torch tensor): input cell image patches
+
+        Returns:
+            dict: generator/discriminator losses
+
+        """
+        # inputs: Batch * num_inputs(channel) * H * W, each channel from 0 to 1
+        z_data = self.enc(inputs)
+        z_prior = t.randn_like(z_data)
+        _z_data = self.enc_d(z_data)
+        _z_prior = self.enc_d(z_prior)
+        g_loss = -t.mean(t.log(_z_data + eps))
+        d_loss = -t.mean(t.log(_z_prior + eps) + t.log(1 - _z_data.detach() + eps))
+        return {'generator_loss': g_loss,
+                'descriminator_loss': d_loss,
+                'score': t.mean(_z_data)}
+
+    def predict(self, inputs):
+        """ Prediction fn, same as forward pass """
+        return self.forward(inputs)
+
+
+
+def train(model, dataset, output_dir, relation_mat=None, mask=None,
+          n_epochs=10, lr=0.001, batch_size=16, gpu=True, shuffle_data=False,
+          transform=None):
+    """ Train function for VQ-VAE, VAE, IWAE, etc.
+
+    Args:
+        model (nn.Module): autoencoder model
+        dataset (TensorDataset): dataset of training inputs
+        relation_mat (scipy csr matrix or None, optional): if given, sparse 
+            matrix of pairwise relations
+        mask (TensorDataset or None, optional): if given, dataset of training 
+            sample weight masks
+        n_epochs (int, optional): number of epochs
+        lr (float, optional): learning rate
+        batch_size (int, optional): batch size
+        gpu (bool, optional): if the model is run on gpu
+        shuffle_data (bool): shuffle data at the end of the epoch to add randomness to mini-batch.
+            Set False when using matching loss
+        transform (bool): data augmentation
+    
+    Returns:
+        nn.Module: trained model
+
+    """
+    optimizer = t.optim.Adam(model.parameters(), lr=lr, betas=(.9, .999))
+    model.zero_grad()
+    n_samples = len(dataset)
+    n_batches = int(np.ceil(n_samples/batch_size))
+    # Declare sample indices and do an initial shuffle
+    sample_ids = np.arange(n_samples)
+    if shuffle_data:
+        np.random.shuffle(sample_ids)
+    writer = SummaryWriter(output_dir)
```

Comment:
Thanks for adding the writer functionality here! This could make logging easy, could you directly update the codes in `vq_vae_supp.py`?

Reply:
Sure. Will try to merge `run_training.py` with `vq_vae_supp.py` in the next PR